Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
     return m_currentOpacity < 1;
 }
 
+bool TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective() const
+{
+    // Check if 3D scene leaf or if has perspective and has children and has 3D transformation
+    if (((!m_state.preserves3D && (m_parent && m_parent->preserves3D())) || m_layerTransforms.localTransform.hasPerspective())
```

Comment:
Does that code exist in a similar way in GraphicsLayerCA & co?

Reply:
nope, CALayer does flattening internally in Core Animation based on given layer properties